### PR TITLE
fix(discover) Literal caching is not typed

### PR DESCRIPTION
**Pull request by evanh**
This actually fixes two bugs. The first is that we were incorrectly caching
Literals with equal int and float values as the same thing. This would break
the subsequent Clickhouse query since the typing wouldn't match.

The second is that the SnQL parser was not correctly handling the HAVING clause.

### CODE REVIEW

File: snuba/query/snql/parser.py [hunk 1]
```diff
@@ -457,9 +457,9 @@ def visit_where_clause(
         return conditions
 
     def visit_having_clause(
-        self, node: Node, visited_children: Tuple[Any, Any, Expression]
+        self, node: Node, visited_children: Tuple[Any, Any, Any, Expression]
     ) -> Expression:
-        _, _, conditions = visited_children
+        _, _, _, conditions = visited_children
```
Comment (fpacifici) [inline snuba/query/snql/parser.py:462]: Is this change actually related to the NOT IN / NOT LIKE change above ?

**Review by fpacifici — APPROVED** [review]
Are the fixes for the the cache and having related in any way? Going on I would advise not to keep unrelated fixes in the same PR (especially if the PR complexity is not small like this one), that throws off reviewers who have a hard time separating them.

File: snuba/query/snql/parser.py (hunk 1 above)
Comment (evanh) [reply snuba/query/snql/parser.py:462]: The changes are linked incidentally. The discover test for the cached failure revealed that we weren't sending `having` clauses through conftest. Fixing that revealed that SnQL was breaking on the HAVING clause. Fixing _that_ revealed a different test that had a NOT IN in the HAVING clause, which we weren't handling correctly.

So while in theory the fixes are independent, I couldn't fix one without fixing the others.

Comment (evanh) [reply snuba/query/snql/parser.py:462]: Yes, indirectly. This is fixing the having clause in general. The NOT IN change fixes a test that breaks once the having clause is fixed: `test_tags_key_boolean_condition` which has a NOT IN in the having clause.

**Comment by evanh** [comment]
> Are the fixes for the the cache and having related in any way? Going on I would advise not to keep unrelated fixes in the same PR (especially if the PR complexity is not small like this one), that throws off reviewers who have a hard time separating them.

The changes are linked incidentally. The discover test for the cached failure revealed that we weren't sending `having` clauses through conftest. Fixing that revealed that SnQL was breaking on the HAVING clause. Fixing _that_ revealed a different test that had a NOT IN in the HAVING clause, which we weren't handling correctly.

So while in theory the fixes are independent, I couldn't fix one without fixing the others.
